### PR TITLE
add: copy code block button to code snippet sections

### LINKED ISSUE #256: Add ability to copy/paste code block
It would be good if there were a button in the top-right corner of the codeblock that allowed users to copy the code with one click of the button.

### PR DESCRIPTION
**Pull request by marksist300**
## Describe your changes

Added a copy and paste button to code snippets. The button's symbol will change on click. It will display "copy" text on hover and focus and "copied" on click. Is aria-labelled so should also be accessible for screen-readers. Have tested pages with Axe Devtools and found no issues.

## Screenshots - If Any (Optional)
copy and paste button for code snippet sections

## Link to issue

Closes #256

## Checklist before requesting a review

- [X] I have performed a self-review of my code.
- [X] Followed the repository's [Contributing Guidelines](/CONTRIBUTING.md).
- [X] I ran the app and tested it locally to verify that it works as expected.
- [X] I have checked my code with an automatic accessibility tool such as Axe Dev Tools or Wave 
      and it shows no errors.
## Additional Information (Optional)
Any additional information that you want to give us.

### CODE REVIEW

**Comment by EmmaDawsonDev** [comment]
Thanks for contributing! I’m currently away on a business trip so I’ll check this properly at the weekend.

**Comment by marksist300** [comment]
> Thanks for contributing! I’m currently away on a business trip so I’ll check this properly at the weekend.

No worries, thanks Emma. I hope you're enjoying the trip! 🙂

File: components/CodeBlock/CodeBlock.tsx [hunk 1]
```diff
@@ -15,11 +16,15 @@ interface ICodeBlockProps {
 
 export const CodeBlock = ({ codeSnippet, languageType }: ICodeBlockProps) => {
 	return (
-		<SyntaxHighlighter
-			language={languageType}
-			style={dracula}
-			wrapLongLines={true}>
-			{codeSnippet}
-		</SyntaxHighlighter>
+		<div style={{ position: "relative" }}>
```
Comment (EmmaDawsonDev) [inline components/CodeBlock/CodeBlock.tsx:19]: Please move all inline styles to a module.css file. If one doesn't yet exist for a component, feel free to make one.

File: components/CopyCodeBlock/CopyCodeBlock.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+	const iconColor = copyOk ? "#0af20a" : "#ddd"
+
+	const handleClick = () => {
+		navigator.clipboard.writeText(code)
+
+		setCopyOk(true)
+		setTimeout(() => {
+			setCopyOk(false)
+		}, 500)
```
Comment (EmmaDawsonDev) [inline components/CopyCodeBlock/CopyCodeBlock.tsx:20]: 500 ms is too fast for some people to see the change and have time to read the text. It's good interaction feedback but I think it needs to last a bit longer so people have time to see it properly.

File: components/CopyCodeBlock/CopyCodeBlock.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import style from "./copyCodeBlock.module.css"
+
+interface CodeSnippet {
+	code: string
+}
+
+const CopyCodeBlock = ({ code }: CodeSnippet) => {
+	const [copyOk, setCopyOk] = useState(false)
+
+	const iconColor = copyOk ? "#0af20a" : "#ddd"
```
Comment (EmmaDawsonDev) [inline components/CopyCodeBlock/CopyCodeBlock.tsx:12]: Please use the color variables found in the global css file.

File: components/CopyCodeBlock/copyCodeBlock.module.css [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/* Styles for the button before being clicked */
+
+.codeCopyBtn {
+	color: white;
```
Comment (EmmaDawsonDev) [inline components/CopyCodeBlock/copyCodeBlock.module.css:4]: White text on a grey background is not readable. Please use an appropriate variable from the global css file

File: components/CopyCodeBlock/copyCodeBlock.module.css [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	cursor: pointer;
+	transition: all 0.3s ease-in-out;
+}
+
+.codeCopyBtn:hover {
+	transform: scale(1.1);
+	opacity: 0.9;
+}
+
+.codeCopyBtn:hover::after {
```
Comment (EmmaDawsonDev) [inline components/CopyCodeBlock/copyCodeBlock.module.css:18]: Is it possible for the copy/copied text to be present on the button all the time? I'd much prefer to have a button with icon and text next to each other where the text is 1rem even if the button is a little bigger. That also makes it easier for people to press.

File: components/CopyCodeBlock/copyCodeBlock.module.css [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+.codeCopyBtn:focus::after {
+	content: "copy";
+	display: flex;
+	align-items: center;
+	font-size: 0.5rem;
+	opacity: 0.9;
+}
+
+/* Styles for button after it has been clicked */
+.codeCopied {
```
Comment (EmmaDawsonDev) [inline components/CopyCodeBlock/copyCodeBlock.module.css:35]: This is a repetition of the codeCopyBtn class so it not needed. They can be combined into one class, especially if the ::after is removed and added as text to the button. Then that can be condidtionally rendered by the state inside the component

File: components/CopyCodeBlock/CopyCodeBlock.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+		}, 500)
+	}
+
+	return (
+		<button
+			className={copyOk ? `${style.codeCopied}` : style.codeCopyBtn}
+			onClick={handleClick}
+			aria-label="Copy code snippet to clipboard">
+			{copyOk ? (
+				<FaCheckSquare style={{ color: iconColor }} />
```
Comment (EmmaDawsonDev) [inline components/CopyCodeBlock/CopyCodeBlock.tsx:29]: Do the icons not have an inbuilt color prop?

**Review by EmmaDawsonDev — CHANGES_REQUESTED** [review]
Hi Mark!

The logic for this works really nicely, there's mainly just some CSS cleanup that's needed before this is ready to merge. The project colors are already set in the global CSS file. Please use those when choosing colors. The current color choices work ok in dark mode but not light mode so be sure to check both versions (if you use the variables most are already set up to work in both light and dark mode).

File: components/CopyCodeBlock/CopyCodeBlock.tsx (hunk 2 above)
Comment (marksist300) [reply components/CopyCodeBlock/CopyCodeBlock.tsx]: You're right, I've changed it to 3 seconds, but can make it longer if you think that'd be better 🙂

File: components/CodeBlock/CodeBlock.tsx (hunk 1 above)
Comment (marksist300) [reply components/CodeBlock/CodeBlock.tsx]: Yep, of course. Have jjust made a new CSS module in the directory

**Comment by marksist300** [comment]
Hi @EmmaDawsonDev. Thank you for the feedback. I've made some changes to the CSS, everything is now based on the root CSS variables, hopefully things are a more readable now. I've added the "copy" / "copied" text to the button so that it's always under the icon. I think everything should be addressed, but let me know if it needs any other changes 🙂. Have a nice weekend!

**Review by EmmaDawsonDev — APPROVED** [review]
This is great, thanks so much for your help!

**Pull request merged**